lib/pure/asyncfutures.nim

```diff
@@ -384,7 +384,7 @@ proc read*[T](future: Future[T] | FutureVar[T]): T =
       injectStacktrace(fut)
       raise fut.error
     when T isnot void:
-      result = fut.value
+      result = move fut.value
```

Comment:
Are the tests green without this breaking change? Also, this should really be `lent T` and you're probably just not understanding that `let x = read()` does still copy because that's what `let x` does.

Reply:
Also red, I'm looking for the cause

Reply:
it doesn't make sense for `read` to move things out - in chronos, this operation is indeed `lent T` - this should be a separate operation such as `drain` because generally, it should be possible to read the future twice and access the same value.. consider code like:

```
let 
  a = myfuture.read().field1
  b = myfuture.read().field2
```

In code that uses chronos, we have seen lots of code like the above "in the wild".

Reply:
If `move` is not used, then the closure will copy a copy of T, resulting in double free. Maybe I should switch to a ref object

```nim
type
  GetAddrInfoResult* = GetAddrInfoResultObj
  GetAddrInfoResultObj = object # wrap uv_getaddrinfo
    res: ptr AddrInfo

proc connect*(s: AsyncSocket, address: string, port: Port) {.async.} =
  var lastExpection: ref Exception = nil
  let addrInfos = await getAddrInfo(s.loop, address, port)
  for address in addrInfos:
    try:
      await connect(s, address.ai_addr)
      return
    except CatchableError as e:
      lastExpection = e
      continue

  raise lastExpection

```

Reply:
> Are the tests green without this breaking change?

yes, is green on `devel` branch.
How can I avoid copying, there will always be a copy inside the closure. I'd like to want to transfer ownership directly to closures